File: labs/lab07/log_probability.py
#!/usr/bin/python

import sys, re, glob, string, os, math



def totalWords(file):
	words = []
	f = open(file, 'r')
	for lines in f:
		lines = lines.lower()
		words += re.split(r'[^a-zA-Z]', lines)
	
	words = filter(None, words)
	return len(words)


def countWords(file, word):
        words = []
        f = open(file, 'r')
        for lines in f:
                lines = lines.lower()
                words += re.split(r'[^a-zA-Z]', lines)
        words = filter(None, words)
	
	found = 0
	for w in words:
		if (word == w):
			found += 1

        return found


def getName(unfiltered):
	new = unfiltered.split("/")
	new = new[1]
	new = new.split(".")
	new = new[0]
	new = new.replace("_", " ")
	return new


for file in sorted(glob.glob("lyrics/*.txt")):
	total = totalWords(file)
	count = countWords(file, sys.argv[1])
	freq = math.log(float(count+1) / (total))
	artist = getName(file)
	print "log(({:d}+1)/{:6d}) = {:.4f} {}".format(count, total, freq, artist)

File: labs/lab07/log_probability.pl
#!/usr/bin/perl

# count words in a file
sub countWords {
	my ($find, $file) = @_;
	$count = 0;	
	open F, $file or die;
        while ($line = <F>) {
                chomp $line;
		$line = lc $line;
                @words = grep(/./, split(/[^a-zA-Z]/, $line));
		# loop through each line and check for the word
        	foreach $i (@words) {
                	if ($i =~ /^$find$/i) {
                        	$count++;
                	}
        	}
        }
	return $count;
}


# counts total words in a file
sub totalWords {
	my ($file) = @_;
	open F, $file or die;
	$count = 0;
	while ($line = <F>) {
		chomp $line;
		@words = grep(/./, split(/[^a-zA-Z]/, $line));
		$count += 0+@words;
	}
	return $count;
}

foreach $file (glob "lyrics/*.txt") {	
	$cw = countWords(@ARGV[0], $file);
	$ncw = $cw + 1;
	$tw = totalWords($file);
	$fq = log($ncw/$tw);
	# get artist name
	@fn = split("/", $file);
	@ufn = split(".txt", @fn[1]);
	$artist = @ufn[0];
	$artist =~ s/_/ /g;
	
	printf("log((%d+1)/%6d) = %8.4f %s\n", $cw, $tw, $fq, $artist);
	
}
